### DISCUSSION TITLE
Radio Calibration problem

### DISCUSSION THREAD

**Discussion by Tjohoansson** [HefnySco/OBAL#44] [Trouble Shooting] [answered]
Hey Mr. Hefny!
We have some problems that we would like some help with..
When we connect to Mission Planner and start Radio Calibration, nothing happens. No green bars are visible at all..
We use a raspberry pi zero 2 w with buster and
Flysky FS-i6X i6X 10CH 2.4GHz AFHDS 2A RC Transmitter With FS-iA10B Receiver
We have measured the inputs on the raspberry with an oscilloscope and there is a signal.
We do not use the GY-91 version.
Do you have any ideas or tips that we can try?
Sincerely.
/ Anders and Albin

**Comment by usbalbin** [comment]
To add some more info

In our last attempt we used the image recommended from [this comment](https://github.com/HefnySco/OBAL/discussions/24#discussioncomment-3831079) and a precompiled [obal 4.2.2]( https://firmware.ardupilot.org/Copter/stable-4.2.2/obal/) binary with the preperations from [this video](https://youtu.be/mYn49GbUL8Y?si=2MsuWBygugF-p1gp) except the compilation part since we used the precompiled binary.

---

In the attempts before that, we ran bullseye and tried to compile from source natively on the pi zero 2 w with the exact same result. All meters "empty" in the radio calibration screen.

![image](https://github.com/HefnySco/OBAL/assets/10927717/391719b2-96f6-48ff-9593-090c7354bd48)

In addition to read the pins with an oscilloscope to confirm the transmitter actually sends pulses to the pi, we also made a small bash script to read the pins. While the scripts is too slow to show the curveform, it does show that the inputs does change.

---

Other than that we have also tried to cross compile ardupilot as is done in [the video](https://youtu.be/mYn49GbUL8Y?si=2MsuWBygugF-p1gp) however running the generated binary on the pi produces a segfault as soon as mission-planner tries to connect. I do not have the link to the exact crosscompiler version used but I believe it was one of [these](https://github.com/abhiTronix/raspberry-pi-cross-compilers)...

---

Is there anything obvious we might be doing wrong? Or something to double check, missing info etc., please let us know :)

/ Albin and Anders

**Reply by HefnySco** [reply]
Building steps are in [https://github.com/HefnySco/OBAL/wiki/06---Build-and-Install](https://github.com/HefnySco/OBAL/wiki/06---Build-and-Install)

**Reply by usbalbin** [reply]
Thanks,

Will give it a try :)

**Reply by Tjohoansson** [reply]
Hi, that did not go as planned

## On my laptop:

```bash
$ git clone https://github.com/ArduPilot/ardupilot.git # Ok
$ cd ardupilot # Ok
$ git submodule update --init  --recursive # Ok
$ make obal # Error, arm-linux-gnueabihf-something not found

[download and unpack the toolchain linked to in wiki]  # Ok

$ ./waf configure --toolchain=/path/where-i-unpacked-the-toolchain/bin/arm-linux-gnueabihf --board obal # Ok
$ ./waf copter # Ok
$ scp /some/path/arducopter pi@some-ip:. # Ok

$ ssh pi@some-ip
pi@raspberrypi:~ $ sudo ./arducopter -A tcp:0.0.0.0:5760:bcast 
RPI 2, 3 or Zero-2
Peripheral base address is 3f000000
Loaded defaults from /home/pi/ardupilot.parm
Config Error: Baro: unable to initialise driver
Config Error: Baro: unable to initialise driver
Config Error: Baro: unable to initialise driver
Config Error: Baro: unable to initialise driver
Config Error: Baro: unable to initialise driver
Config Error: Baro: unable to initialise driver
Config Error: Baro: unable to initialise driver
Config Error: Baro: unable to initialise driver
Config Error: Baro: unable to initialise driver
Interrupted: Segmentation fault                           # <--------- This is when connecting with mission planner
pi@raspberrypi:~ $
```

Mission planner does show the version to be 4.5.0-dev before disconnecting due to the segfault
The segfaut does not appear when using the precompiled binaries(4.2.2 or 4.4.0). However the barometer problem appears on both precompiled and self built versions.

## On the raspberry with buster
```bash
sudo apt update
sudo apt install git
$ git clone https://github.com/ArduPilot/ardupilot.git # Ok
$ cd ardupilot # Ok
$ git submodule update --init  --recursive # Ok

$ make obal # Note that python3 is installed, but it seems to check `python` which is python2
Setting top to                           : /home/pi/code/ardupilot
Setting out to                           : /home/pi/code/ardupilot/build
Autoconfiguration                        : enabled
Checking for program 'python'            : /usr/bin/python
Checking for python version >= 3.6.9     : 2.7.16
The python version is too old, expecting (3, 6, 9)
(complete log in /home/pi/code/ardupilot/build/config.log)
make: *** [Makefile:31: obal] Error 1
$ sudo apt upgrade
$ make obal # same problem
```

What are we doing wrong?

**Reply by HefnySco** [reply]
Normally I use 
`$sudo /home/pi/arducopter -A udp:192.168.1.123:14550 -B /dev/serial0 -C serial1
`
 instead of 
`$ sudo ./arducopter -A tcp:0.0.0.0:5760:bcast 
`
[https://github.com/HefnySco/OBAL/wiki/06---Build-and-Install](https://github.com/HefnySco/OBAL/wiki/06---Build-and-Install)

I am worried of "Config Error: Baro: unable to initialise driver".
If you sure that SPI is active then I am afraid that you use a [fake GP-91](https://github.com/HefnySco/OBAL/wiki/07---OBAL-GY91). I bought one from aliexpress that I discovered that the IC reply with with another ID and it was not what it should to be.

**Reply by usbalbin** [reply]
We do *not* use the GY-91 variant of the board

**Reply by HefnySco** [reply]
so you are using the original variant that uses IMU-9250 with external barometer.

could you specify the exact sensors you are using? and please make sure that i2c & spi are enabled from config.

**Reply by Tjohoansson** [reply]
Good morning!

We have built the original version and used the cards you specified:
16 Channel 12-bit PWM/Servo Driver-I2C interface PCA9685
16 Bit I2C ADS1115 Module ADC 4 channel with Pro Gain Amplifier RPi
GY-68 BMP180 Replace BMP085 Digital Barometric Pressure Sensor Module
MPU-9250 GY-9250 9-axis sensor module I2C/SPI Communications Thriaxis gyroscope

We will check the  i2c & spi configuration!

Thanks for your help!

**Reply by Tjohoansson** [reply]
Good morning!
Now we have both checked i2c and spi without results..
We will solder a new board to check if it is some of the sensor card that is not working..
We will never give up!

**Reply by HefnySco** [reply]
Please use python code to run it on your RPI and test the sensors while soldered on the PCB (OBAL). This will confirm if soldering and sensors are good or not.
If sensors are working and ardupilot firmware is running then it should work fine. but first confim using python test code running on RPI.

**Reply by usbalbin** [reply]
Thanks!

Any hints on where too start looking for such code?

**Reply by HefnySco** [reply]
Please note that Bulleye64 Need to be **compiled** to enable **CONFIG_STRICT_DEVMEM** check this [link ](https://github.com/HefnySco/OBAL/wiki/05---Configuring-Raspberry).
Also please compile ardupilot master code as there is a bug in the last release. The newer release includes the fix already.

**Comment by usbalbin** [comment]
Measuring on GPIO5 (CH1) on the raspberry, as far as we can see, everything looks as it should.

![image](https://github.com/HefnySco/OBAL/assets/10927717/c853b5d1-eb8d-4502-8b7a-1f4d2d8223e4)

And the pulse width changes from 1-2ms depending on stick position on the controller.

**Comment by HefnySco** [comment]
Hello,

1- Please run OBAL from command prompt and make sure that there is no problems in reading the sensors (IMPORTANT)

2- Are you using PPM,PWM ?

Also please make sure you compile master version of ardupilot not release. as I believe this fix is not merged yet
[https://github.com/ArduPilot/ardupilot/pull/23943
](https://github.com/ArduPilot/ardupilot/pull/23943
)

**Reply by usbalbin** [reply]
Thanks a lot for the answer :)

We are using PWM, the remote and receiver support both, and we tried both settings with no difference. We also verified the settings with the oscilloscope. With that said, PWM was the plan.

I do not have a screenshot but there were warnings about the barometer

**Reply by Tjohoansson** [reply, marked as answer]
Good news Mr. Hefny!
Our barometer card was broken!
Now we have changed and everything seems to work fine!
Finally!
